distributed/joblib.py

```diff
@@ -116,47 +174,80 @@ def get_nested_backend(self):
     def configure(self, n_jobs=1, parallel=None, **backend_args):
         return self.effective_n_jobs(n_jobs)
 
+    def start_call(self):
+        self.call_data_futures = _WeakKeyDictionary()
+
+    def stop_call(self):
+        # The explicit call to clear is required to break a cycling reference
+        # to the futures.
+        self.call_data_futures.clear()
+        self.call_data_futures = None
+
     def effective_n_jobs(self, n_jobs):
         return sum(self.client.ncores().values())
 
     def _to_func_args(self, func):
-        if not self.data_to_future:
-            return func, ()
-        args2 = []
-        lookup = dict(self.data_to_future)
+        collected_futures = []
+        itemgetters = dict()
+
+        # Futures that are dynamically generated during a single call to
+        # Parallel.__call__.
+        call_data_futures = getattr(self, 'call_data_futures', None)
 
         def maybe_to_futures(args):
-            for x in args:
-                id_x = id(x)
-                if id_x in lookup:
-                    x = lookup[id_x]
-                    if type(x) is not itemgetter:
-                        x = itemgetter(len(args2))
-                        args2.append(lookup[id_x])
-                        lookup[id_x] = x
-                yield x
+            for arg in args:
+                arg_id = id(arg)
+                if arg_id in itemgetters:
+                    yield itemgetters[arg_id]
+                    continue
+
+                f = None
+                if f is None and arg_id in self.data_futures:
+                    f = self.data_futures[arg_id]
+
+                if f is None and call_data_futures is not None:
+                    try:
+                        f = call_data_futures[arg]
+                    except KeyError:
+                        if (call_data_futures is not None
```

Comment:
`call_data_futures` will always be not `None` due to the `if` check above, this check is not needed.